Bucket listing is to eager

# 🐛 Description

A couple of times a day we get this problem when processing Level1A:

```
[ERROR] Level1AException: Failed to get aux data for 2023/2/21/23/MATS_OPS_Level0_VC1_APID100_20230221-132121_20230222-132833.parquet with start time 2023-02-21 22:59:55.126846313+00:00 and end time 2023-02-21 23:59:55.410049438+00:00:
  When listing objects under key 'HTR' in bucket 'ops-payload-level0-v0.2':
    AWS Error SLOW_DOWN during ListObjectsV2 operation:
      Please reduce your request rate.
```

The problem started when we shifted to hourly partitioning and will, unfortunately, only get worse as we get on...


# ✌🏽 Expected behaviour

Files pass  through without issues.

--- a/level1a/handlers/level1a.py
+++ b/level1a/handlers/level1a.py
@@ -1,6 +1,7 @@
 import json
 import os
 from http import HTTPStatus
+from time import sleep
 from traceback import format_tb
 from typing import Any, Callable, Dict, Optional, Tuple
 
@@ -35,6 +36,7 @@ except ImportError:
 Event = Dict[str, Any]
 Context = Any
 
+RETRIES = 5
 OFFSET_FACTOR = 2
 RECONSTRUCTED_FREQUENCY = 1.  # Hz
 HTR_FREQUENCY = 0.1  # Hz
@@ -70,14 +72,21 @@ class Level1AException(Exception):
     pass
 
 
-def s3_backoff(caller: Callable):
-    def inner(*args, **kwargs):
-        try:
-            caller(*args, **kwargs)
-        except Exception:
-            pass
+class RetriesExceeded(Exception):
+    pass
 
-    return inner
+
+def s3_backoff(caller: Callable):
+    def wrapper(*args, **kwargs):
+        msg = ""
+        for r in range(RETRIES + 1):
+            try:
+                sleep(2 ** r - 1)
+                return caller(*args, **kwargs)
+            except Exception as err:
+                msg = str(err)
+        raise RetriesExceeded(msg)
+    return wrapper
 
 
 def get_or_raise(variable_name: str) -> str:
@@ -395,7 +404,7 @@ def lambda_handler(event: Event, context: Context):
         min_time, max_time = get_search_bounds(rac_df.index)
     except Exception as err:
         tb = '|'.join(format_tb(err.__traceback__)).replace('\n', ';')
-        msg = f"Failed to initialize handler: {err} ({tb})"
+        msg = f"Failed to initialize handler: {err} ({type(err)}; {tb})"
         raise Level1AException(msg)
 
     try:

--- a/level1a/handlers/level1a.py
+++ b/level1a/handlers/level1a.py
@@ -119,6 +119,7 @@ def covers(
     )
 
 
+@s3_backoff
 def get_level0_records(
     path_or_bucket: str,
     filesystem: pa.fs.FileSystem = None,
@@ -134,6 +135,7 @@ def get_level0_records(
     )
 
 
+@s3_backoff
 def get_htr_records(
     path_or_bucket: str,
     min_time: Timestamp,
@@ -195,6 +197,7 @@ def get_htr_records(
     return dataset
 
 
+@s3_backoff
 def get_reconstructed_records(
     path_or_bucket: str,
     min_time: Timestamp,

--- a/level1a/handlers/level1a.py
+++ b/level1a/handlers/level1a.py
@@ -1,5 +1,6 @@
 import json
 import os
+from functools import wraps
 from http import HTTPStatus
 from time import sleep
 from traceback import format_tb
@@ -77,6 +78,7 @@ class RetriesExceeded(Exception):
 
 
 def s3_backoff(caller: Callable):
+    @wraps(caller)
     def wrapper(*args, **kwargs):
         msg = ""
         for r in range(RETRIES + 1):

--- a/level1a/handlers/level1a.py
+++ b/level1a/handlers/level1a.py
@@ -2,7 +2,7 @@ import json
 import os
 from http import HTTPStatus
 from traceback import format_tb
-from typing import Any, Dict, Optional, Tuple
+from typing import Any, Callable, Dict, Optional, Tuple
 
 import numpy as np
 import pyarrow as pa  # type: ignore
@@ -70,6 +70,16 @@ class Level1AException(Exception):
     pass
 
 
+def s3_backoff(caller: Callable):
+    def inner(*args, **kwargs):
+        try:
+            caller(*args, **kwargs)
+        except Exception:
+            pass
+
+    return inner
+
+
 def get_or_raise(variable_name: str) -> str:
     if (var := os.environ.get(variable_name)) is None:
         raise EnvironmentError(
@@ -100,16 +110,17 @@ def covers(
     )
 
 
-def get_ccd_records(
+def get_level0_records(
     path_or_bucket: str,
     filesystem: pa.fs.FileSystem = None,
+    index: str = "EXPDate",
 ) -> Tuple[DataFrame, pq.FileMetaData]:
     table = pq.read_table(
         path_or_bucket,
         filesystem=filesystem,
     )
     return (
-        table.to_pandas().set_index("EXPDate").sort_index(),
+        table.to_pandas().set_index(index).sort_index(),
         table.schema.metadata,
     )
 
@@ -370,7 +381,7 @@ def lambda_handler(event: Event, context: Context):
                 })
             }
 
-        rac_df, metadata = get_ccd_records(
+        rac_df, metadata = get_level0_records(
             f"{bucket}/{object_path}",
             filesystem=s3,
         )
@@ -394,27 +405,28 @@ def lambda_handler(event: Event, context: Context):
             max_time + get_offset(RECONSTRUCTED_FREQUENCY),
             filesystem=s3,
         )
-        htr_df = get_htr_records(
-            f"{htr_bucket}/HTR",
-            min_time - get_offset(HTR_FREQUENCY),
-            max_time + get_offset(HTR_FREQUENCY),
-            filesystem=s3,
-        )
 
         if not covers(reconstructed_df.index, min_time, max_time):
             raise DoesNotCover("Reconstructed data is missing timestamps")
-    except Exception as err:
-        tb = '|'.join(format_tb(err.__traceback__)).replace('\n', ';')
-        msg = f"Failed to get aux data for {output_path} with start time {min_time} and end time {max_time}: {err} ({tb})"  # noqa: E501
-        raise Level1AException(msg)
 
-    try:
         reconstructed_df = interpolate(
             reconstructed_df,
             rac_df.index,
             max_diff=get_offset(RECONSTRUCTED_FREQUENCY),
         )
         reconstructed_df = add_satellite_position_data(reconstructed_df)
+    except Exception as err:
+        tb = '|'.join(format_tb(err.__traceback__)).replace('\n', ';')
+        msg = f"Failed to get reconstructed data for {output_path} with start time {min_time} and end time {max_time}: {err} ({type(err)}; {tb})"  # noqa: E501
+        raise Level1AException(msg)
+
+    try:
+        htr_df = get_htr_records(
+            f"{htr_bucket}/HTR",
+            min_time - get_offset(HTR_FREQUENCY),
+            max_time + get_offset(HTR_FREQUENCY),
+            filesystem=s3,
+        )
         htr_subset = interpolate(
             htr_df,
             rac_df.index,
@@ -422,7 +434,7 @@ def lambda_handler(event: Event, context: Context):
         )
     except Exception as err:
         tb = '|'.join(format_tb(err.__traceback__)).replace('\n', ';')
-        msg = f"Failed to transform aux data for {output_path} with start time {min_time} and end time {max_time}: {err} ({tb})"  # noqa: E501
+        msg = f"Failed to get HTR data for {output_path} with start time {min_time} and end time {max_time}: {err} ({type(err)}; {tb})"  # noqa: E501
         raise Level1AException(msg)
 
     try:
@@ -443,5 +455,5 @@ def lambda_handler(event: Event, context: Context):
         )
     except Exception as err:
         tb = '|'.join(format_tb(err.__traceback__)).replace('\n', ';')
-        msg = f"Failed to store {output_path} with start time {min_time} and end time {max_time}: {err} ({tb})"  # noqa: E501
+        msg = f"Failed to store {output_path} with start time {min_time} and end time {max_time}: {err} ({type(err)}; {tb})"  # noqa: E501
         raise Level1AException(msg)

--- a/tests/level1a/handlers/test_level1a.py
+++ b/tests/level1a/handlers/test_level1a.py
@@ -11,7 +11,7 @@ import pytest  # type: ignore
 from level1a.handlers.level1a import (
     HTR_COLUMNS,
     covers,
-    get_ccd_records,
+    get_level0_records,
     get_htr_records,
     get_or_raise,
     get_reconstructed_records,
@@ -96,8 +96,8 @@ def test_covers(indices, first, last, expect):
     assert covers(indices, first, last) == expect
 
 
-def test_get_ccd_records(ccd_path):
-    out, meta = get_ccd_records(ccd_path)
+def test_get_level0_records(ccd_path):
+    out, meta = get_level0_records(ccd_path)
     expect_inds = pd.DatetimeIndex(
         [
             '2022-12-21 23:59:59.063491821+00:00',

